File: android/app/src/main/java/com/seoyoung/withu/shared/WatchMessage.kt
package com.seoyoung.withu.shared

import com.seoyoung.withu.character.CharacterState
import kotlinx.serialization.Serializable

/**
 * "현재 적용 스냅샷" — iOS WatchMessage 포팅 (스펙 09 §3-10).
 * Android 에선 Watch 전송은 없지만 홈/위젯(Glance)이 같은 데이터를 읽는
 * 단일 소스 역할은 동일 — 파리티를 위해 이름 유지.
 * 필수는 state/timestamp 뿐 — 나머지는 옛 메시지 디코드 호환을 위해 전부 nullable.
 */
@Serializable
data class WatchMessage(
    val state: String,                       // CharacterState.raw
    val todaySteps: Double? = null,
    val lastSleepHours: Double? = null,
    val todayActiveMinutes: Double? = null,  // 오늘 활동(운동) 분
    val todayActiveKcal: Double? = null,     // 오늘 활성 칼로리
    val weatherEmoji: String? = null,        // 예: ☀️
    val weatherTempC: Double? = null,        // 섭씨
    val weatherSunrise: Long? = null,        // epoch millis
    val weatherSunset: Long? = null,         // epoch millis
    val timestamp: Long,
) {
    /** raw 가 알 수 없는 값이어도 깨지지 않게 IDLE fallback. */
    val characterState: CharacterState
        get() = CharacterState.fromRaw(state) ?: CharacterState.IDLE
}

File: withu/Shared/WatchMessage.swift
//
//  WatchMessage.swift
//  withu (Shared between iOS and watchOS targets)
//
//  Xcode UI 에서 이 파일의 Target Membership 을 iOS + watchOS 둘 다 체크해야 함.
//

import Foundation

/// iPhone → Watch 로 보내는 캐릭터 스냅샷 메시지.
/// `updateApplicationContext` 로 직렬화돼 전달됨.
struct WatchMessage: Codable, Equatable, Sendable {
    let state: CharacterState
    let todaySteps: Double?
    let lastSleepHours: Double?
    /// 오늘 활동(운동) 분
    let todayActiveMinutes: Double?
    /// 오늘 활성 칼로리
    let todayActiveKcal: Double?
    /// 현재 날씨 이모지 (예: ☀️)
    let weatherEmoji: String?
    /// 현재 기온 (섭씨)
    let weatherTempC: Double?
    /// 오늘 일출 (위치 local time). 옵셔널.
    var weatherSunrise: Date? = nil
    /// 오늘 일몰 (위치 local time). 옵셔널.
    var weatherSunset: Date? = nil
    /// 프로필 시간 창 요약 — 워치 컴플리케이션이 미래 수면/기상 전환을 스스로 계산하게 함께 전송.
    var schedule: SharedAppState.ScheduleInfo? = nil
    let timestamp: Date

    /// 직렬화/역직렬화 키. 양쪽이 같은 이름을 쓰게 명시.
    static let payloadKey = "withu.watchMessage.v1"

    // 옛 메시지와 호환: 새 필드는 옵셔널, 디코드 시 nil 허용
    init(state: CharacterState,
         todaySteps: Double? = nil,
         lastSleepHours: Double? = nil,
         todayActiveMinutes: Double? = nil,
         todayActiveKcal: Double? = nil,
         weatherEmoji: String? = nil,
         weatherTempC: Double? = nil,
         weatherSunrise: Date? = nil,
         weatherSunset: Date? = nil,
         schedule: SharedAppState.ScheduleInfo? = nil,
         timestamp: Date = Date()) {
        self.state = state
        self.todaySteps = todaySteps
        self.lastSleepHours = lastSleepHours
        self.todayActiveMinutes = todayActiveMinutes
        self.todayActiveKcal = todayActiveKcal
        self.weatherEmoji = weatherEmoji
        self.weatherTempC = weatherTempC
        self.weatherSunrise = weatherSunrise
        self.weatherSunset = weatherSunset
        self.schedule = schedule
        self.timestamp = timestamp
    }
}
